### PR TITLE
Trait beginnings

### PR DESCRIPTION
**Pull request by flodiebold**
This adds some very simple trait method resolution, going through traits in scope, looking for methods of the given name, and checking very naively whether there's an impl for the given type and trait.

### CODE REVIEW

File: crates/ra_hir/src/resolve.rs [hunk 1]
```diff
@@ -175,6 +175,21 @@ impl Resolver {
...
+            .iter()
+            .rev()
+            .flat_map(|scope| {
+                match scope {
+                    Scope::ModuleScope(m) => Some(m.crate_def_map[m.module_id].scope.traits()),
+                    _ => None,
+                }
+                .into_iter()
+            })
+            .flat_map(|i| i)
```
Comment (matklad) [inline crates/ra_hir/src/resolve.rs]: There's `.flatten`

File: crates/ra_hir/src/ty/method_resolution.rs [hunk 2]
```diff
@@ -109,6 +108,24 @@ impl CrateImplBlocks {
...
+/// will actually be done by chalk.
+pub(crate) fn implements(db: &impl HirDatabase, trait_ref: TraitRef) -> bool {
+    // FIXME use all trait impls in the whole crate graph
+    let krate = trait_ref.trait_.module(db).krate(db);
+    let krate = match krate {
+        Some(krate) => krate,
+        None => return false,
+    };
+    let crate_impl_blocks = db.impls_in_crate(krate);
+    for impl_block in crate_impl_blocks.lookup_impl_blocks_for_trait(&trait_ref.trait_) {
```
Comment (matklad) [inline crates/ra_hir/src/ty/method_resolution.rs]: this could use `.any`

**Comment by flodiebold** [comment]
@matklad I think this can be merged, any further comments?

**Comment by matklad** [comment]
Only

bors r+

:)

**Pull request merged**